antsibull/cli/doc_commands/stable.py

```diff
@@ -221,7 +222,7 @@ def get_plugin_contents(plugin_info: t.Mapping[str, t.Mapping[str, t.Any]],
 
 
 def get_collection_contents(plugin_content: t.Mapping[str, t.Mapping[str, t.Mapping[str, str]]],
-                            ) -> t.DefaultDict[str, t.Dict[str, t.Mapping[str, str]]]:
+                            ) -> CollectionInfoT:
```

Comment:
The reason not to do this would be that the types for return values should be concrete: We know we're returning a DefaultDict of Dicts because we're creating those inside of this function.  So anything making use of our return values would know they can depend on those behaviours being present.  Typing it as generic Mapping of Mapping means that code using our return value can't depend on those behaviours.

Input types, otoh, should specify the least common denominator that they need.  So in terms of taking input, a Mapping is appropriate unless the function wanted to do something that depended on the DefaultDict's behaviour.

I kind of feel like we should keep return types and input types separate for that reason (when they're different).  Does it make sense to define a return spec alongside the input spec so that it's easier to write this sort of thing and see that they're compatible types?  (like `RetCollectionInfoT` as a counterpart to `CollectionInfoT`?)  Or does that defeat the purpose of keeping the input type as lowest common denominator?

Reply:
Usually things like "this implementation uses a `defaultdict`" is an implementation detail and callers should not depend on that, because it makes it really hard to change the implementation without having to go through all callers to make sure they don't depend on it. (If a generic Mapping is documented, they can of course still use it like a `defaultdict`, but then they're clearly faulty.)

Using `CollectionInfoT` here says to API users that "this collection returns what can be used in other APIs as `CollectionInfoT`", as opposed to "we return this concrete mapping type, which might be similar to one of the input types of other functions, but you have to figure that out by yourself". Also, defining `RetCollectionInfoT` different from `CollectionInfoT` makes it impossible to ever provide any function returning that which does not use a similar implementation (based on `defaultdict`).

Reply:
Changed in 54ed7ceedd120d1968a028f733ff269101eab08e.

Reply:
Interesting.....   So I don't think of static typing info as part of the API but I can definitely see your point that if it is, then we should be using it to define capabilities to expect rather than the specific type.  I guess there's two separate use cases (static typing to catch errors in the current implementation vs static typing to define the expected API) and in this specific case they're a little bit at odds.

There are cases where defaultdict and dict behaviour would be different even though you are only using functions that are common to both, so I think it can be beneficial to know concretely what we're returning.  To be fair, I don't think static typing will prevent it (at this stage in type checkers' development, at least), though so type checking might not be the place to specify this and enforce it....  An example would be the following code:

``` python
def test() -> defauldict:
    return defaultdict(int)

d = test()
try:
    print(d['x'])
except KeyError:
    print("d['x'] is unknown")
```